### PR TITLE
Update PTL FW to execute test apart from tests dir

### PR DESCRIPTION
**Pull request by shilpakodli**
#### Describe Bug or Feature
Currently in PTL, to run a test case using "pbs_benchpress -t" option, user has to go default test directory and then run the test case. If user is in non test directory then PTL used to throw an error saying " Unknown testsuite/Unknown testcase"

#### Describe Your Change
PTL will now search tests in current directory first, if it doesn't find test in current directory then it will search in PTL installed tests directory. If it doesn't find in both of the paths then it will search in environment variable "PTLTESTS" where user can explictly mention the path to tests directory.

#### Logs of test execution
[smoktest_results.txt](https://github.com/PBSPro/pbspro/files/3290142/smoktest_results.txt)

### CODE REVIEW

**Comment by agrawalravi90** [comment]
Hi @shilpakodli , thanks for proposing this change, this has irked me for so long!
But I do think that this deserves a design document for 2 reasons:
1) So that a wider audience gets to know that this is finally happening and
2) You are introducing a new environment variable, and you've taken some design decisions for how the tests will be found etc., i think it'll be useful to take feedback from others on this.

Thanks again for doing this

File: test/fw/ptl/utils/plugins/ptl_test_loader.py [hunk 1]
```diff
@@ -191,7 +194,30 @@ def prepareTestLoader(self, loader):
 
         def check_loadTestsFromNames(names, module=None):
             rv = old_loadTestsFromNames(names, module)
-            self.check_unknown()
+            test_list = self.__tests_list_copy
+            test_suites = self.__tests_list_copy['__only__ts__']
+            if (len(test_list) > 2) or (len(test_suites) != 0):
+                if 'PBSTestSuite' not in test_suites:
```
Comment (hirenvadalia) [inline test/fw/ptl/utils/plugins/ptl_test_loader.py]: why do you need this if condition? what if the user is trying to run `pbs_benchpress -i -t PBSTestSuite` when PTL_TESTS_DIR is set to a valid non-install directory

File: test/fw/ptl/utils/plugins/ptl_test_loader.py [hunk 2]
```diff
@@ -191,7 +194,30 @@ def prepareTestLoader(self, loader):
...
+                            rv = old_loadTestsFromNames([user_test_dir],
+                                                        module)
+                        else:
+                            logging.error("Invalid dircetory specified in
+                                          PTL_TESTS_DIR")
+                            sys.exit(1)
+                    elif os.path.isdir(ptl_test_dir):
+                        rv = old_loadTestsFromNames([ptl_test_dir], module)
+                    if not self.__need_info:
+                        self.check_unknown()
```
Comment (hirenvadalia) [inline test/fw/ptl/utils/plugins/ptl_test_loader.py]: What will be an error message if the user runs `pbs_benchpress -i -t UnknownTestSuite` with your changes?

File: test/fw/ptl/utils/plugins/ptl_test_loader.py (hunk 1 above)
Comment (shilpakodli) [reply test/fw/ptl/utils/plugins/ptl_test_loader.py]: When pbs_benchpress is run with -f option where file is not present in current directory then, self.__test_list_copy will have PBSTestSuite in self.__tests_list_copy['__only__ts__'] and it will go in if condition and try to run all of the test suites which is not expected. I have put this condition to skip for -f option.

File: test/fw/ptl/utils/plugins/ptl_test_loader.py (hunk 2 above)
Comment (shilpakodli) [reply test/fw/ptl/utils/plugins/ptl_test_loader.py]: So, in pbs_benchpress, I am calling loader plugin and then testinfo plugin. Loader will load all of the test cases in current/User specified/Installed directory and runs check_unknown function which I don't want to call for -i option as it is taken care when testinfo plugin is called.

File: test/fw/ptl/utils/plugins/ptl_test_loader.py [hunk 3]
```diff
@@ -191,7 +194,30 @@ def prepareTestLoader(self, loader):
 
         def check_loadTestsFromNames(names, module=None):
             rv = old_loadTestsFromNames(names, module)
-            self.check_unknown()
+            test_list = self.__tests_list_copy
+            test_suites = self.__tests_list_copy['__only__ts__']
+            if (len(test_list) > 2) or (len(test_suites) != 0):
```
[the remark this answers is not in the document]
Comment (shilpakodli) [reply test/fw/ptl/utils/plugins/ptl_test_loader.py]: "old_loadTestsFromNames(names, module)" is called at the beginning of check_loadTestsFromNames method with names = "Current Dircetory".
If the given test case/test suite to be run is not present in current directory then "old_loadTestsFromNames" will update self.__tests_list_copy with non-existing test suite/test case.
self.__test_list_copy is assigned initially to "{self._only_ts: [], self._only_tc: []}" when loader plugin is called.
> Suppose if "MytestSuite" which I want to run is not present in current directory then self__test_list_copy will have "{self._only_ts: ["MytestSuite"], self._only_tc: []}"
> If "MytestSuite.testcase1" which is a single test case in MytestSuite test suite is not present in current directory, then self.__test_list_copy will have "{self._only_ts: [], self._only_tc: [], " MytestSuite" : ["testcase1"]}
So "if (len(test_list) > 2) or (len(test_suites) != 0):" is for above two scenarios.

File: test/fw/ptl/utils/plugins/ptl_test_loader.py (hunk 1 above)
Comment (shilpakodli) [reply test/fw/ptl/utils/plugins/ptl_test_loader.py]: When I run this below command 
"pbs_benchpress -i -t PBSTestSuite" , it is giving output of all methods present in ptl/utils/pbs_testsuite.py

File: test/fw/ptl/utils/plugins/ptl_test_loader.py [hunk 4]
```diff
@@ -191,7 +194,31 @@ def prepareTestLoader(self, loader):
 
         def check_loadTestsFromNames(names, module=None):
             rv = old_loadTestsFromNames(names, module)
-            self.check_unknown()
+            test_list = self.__tests_list_copy
```
Comment (hirenvadalia) [inline test/fw/ptl/utils/plugins/ptl_test_loader.py]: @shilpakodli As per our offline discussion, we don't need this change we just need to pass a new directory in `old_loadTestsFromNames` along with names. Also, you won't need a new argument to `set_data` after new changes

**Comment by shilpakodli** [comment]
@hirenvadalia Thanks for the review. I have addressed your offline review comments and tested on my VM.

**Comment by shilpakodli** [comment]
> Hi @shilpakodli , thanks for proposing this change, this has irked me for so long!
> But I do think that this deserves a design document for 2 reasons:
> 
> 1. So that a wider audience gets to know that this is finally happening and
> 2. You are introducing a new environment variable, and you've taken some design decisions for how the tests will be found etc., i think it'll be useful to take feedback from others on this.
> 
> Thanks again for doing this

EDD got sign off now and code is ready for review. Thanks for bringing up this @agrawalravi90

File: test/fw/ptl/utils/plugins/ptl_test_loader.py [hunk 5]
```diff
@@ -190,9 +193,35 @@ def prepareTestLoader(self, loader):
...
 
         def check_loadTestsFromNames(names, module=None):
-            rv = old_loadTestsFromNames(names, module)
+            if not self.testfiles:
+                self._du = DshUtils()
+                ptl_test_dir = self._du.which(exe='pbs_benchpress')
+                ptl_test_dir = ptl_test_dir.replace('bin/pbs_benchpress',
+                                                    'tests')
+                user_test_dir = os.environ.get("PTL_TESTS_DIR", None)
+                if user_test_dir is not None:
```
Comment (hirenvadalia) [inline test/fw/ptl/utils/plugins/ptl_test_loader.py]: as discussed offline please remove this if-else condition and create a new list with names and append ptl_test_dir and user_test_dir based on availability and in order.

**Comment by shilpakodli** [comment]
@hirenvadalia  As per offline discussion, I have addressed your review comments.

**Comment by shilpakodli** [comment]
@hirenvadalia I am attaching test execution results with following observation.
test case : pbs_test_pbs.py
Same test case is present in PTL_TESTS_DIR, PTL Installed Directory and current directory but with different print statement.

PTL_TESTS_DIR (/tmp/tests) - Executed from PTL TESTS Directory
PTL Installed Dir (/opt/ptl/tests) - Executed from PTL Installed Directory
Current Directory (/home/pbsroot/usertemp/tests) - Executed from Current Directory

I have attached logs with three scenarios with condition as same test suite present in all three directories:

scenario1 : Ran from Non-tests directory and PTL_TESTS_DIR set to /tmp/tests/
            Result : Test should execute from PTL_TESTS_DIR.
Scenario2 : Ran from current dircetory and PTL_TESTS_DIR set to /tmp/tests/
            Result : Test should execute from current directory
Scenario3 : Ran from non-tests dircetory and unset PTL_TESTS_DIR
            Result : Test should execute from PTL Installed directory.

[scenario1_results.txt](https://github.com/PBSPro/pbspro/files/3381686/scenario1_results.txt)
[scenario2_results.txt](https://github.com/PBSPro/pbspro/files/3381687/scenario2_results.txt)
[scenario3_results.txt](https://github.com/PBSPro/pbspro/files/3381688/scenario3_results.txt)

File: test/fw/ptl/utils/plugins/ptl_test_loader.py [hunk 6]
```diff
@@ -190,7 +195,20 @@ def prepareTestLoader(self, loader):
         old_loadTestsFromNames = loader.loadTestsFromNames
 
         def check_loadTestsFromNames(names, module=None):
-            rv = old_loadTestsFromNames(names, module)
+            tests_dir = names
+            if not self.testfiles:
+                self._du = DshUtils()
+                ptl_test_dir = self._du.which(exe='pbs_benchpress')
```
Comment (hirenvadalia) [inline test/fw/ptl/utils/plugins/ptl_test_loader.py]: what if pbs_benchpress is not in the path? you can use sys.argv[0].

Comment (shilpakodli) [reply test/fw/ptl/utils/plugins/ptl_test_loader.py]: Can we use PBS_EXEC path and then replace pbs with ptl ?

Comment (hirenvadalia) [reply test/fw/ptl/utils/plugins/ptl_test_loader.py]: I don't think use PBS_EXEC is good. it's not guaranteed that use has installed PTL in /opt/ptl.
So it is best to use sys.argv[0]

Comment (shilpakodli) [reply test/fw/ptl/utils/plugins/ptl_test_loader.py]: I agree. Wish we could have something called PTL_HOME like PBS_HOME which represents PTL installed path. 
For now, I have changed to sys.argv[0] considering user will always use full path of pbs_benchpress to run the test case.

File: test/fw/bin/pbs_benchpress [hunk 7]
```diff
@@ -434,10 +434,13 @@ if __name__ == '__main__':
         follow = True
 
     if only_info:
+        loader = PTLTestLoader()
```
Comment (prakashcv13) [inline test/fw/bin/pbs_benchpress]: we can move this out of if-elif-else block

File: test/fw/ptl/utils/plugins/ptl_test_loader.py [hunk 8]
```diff
@@ -66,14 +68,15 @@ def __init__(self):
...
+        self.testfiles = None
 
     def options(self, parser, env):
         """
         Register command line options
         """
         pass
 
-    def set_data(self, testgroup, suites, excludes, follow):
+    def set_data(self, testgroup, suites, excludes, follow, testfiles=None):
```
Comment (prakashcv13) [inline test/fw/ptl/utils/plugins/ptl_test_loader.py]: add param testfiles in the docstring

**Comment by shilpakodli** [comment]
@hirenvadalia Addressed offline review comments.

**Comment by shilpakodli** [comment]
@prakashcv13 I have addressed your review comments. Please have a look.

File: test/fw/ptl/utils/plugins/ptl_test_loader.py [hunk 9]
```diff
@@ -202,7 +203,8 @@ def check_loadTestsFromNames(names, module=None):
                                             "ptl", "tests")
                 user_test_dir = os.environ.get("PTL_TESTS_DIR", None)
                 if user_test_dir is not None:
-                    tests_dir += [user_test_dir]
+                    if os.path.isdir(user_test_dir):
```
Comment (hirenvadalia) [inline test/fw/ptl/utils/plugins/ptl_test_loader.py]: you can merge this if with above if: `if user_test_dir and os.path.isdir(user_test_dir):`

File: test/fw/ptl/utils/plugins/ptl_test_info.py [hunk 10]
```diff
@@ -128,26 +128,9 @@ def wantClass(self, cls):
         """
         Is the class wanted?
         """
-        if not issubclass(cls, PBSTestSuite):
```
Comment (agrawalravi90) [inline test/fw/ptl/utils/plugins/ptl_test_info.py]: why did you remove this?

File: test/fw/ptl/utils/plugins/ptl_test_info.py [hunk 11]
```diff
@@ -128,26 +128,9 @@ def wantClass(self, cls):
...
-            return False
-        if cls.__name__ == 'PBSTestSuite':
-            return False
         self._tree.setdefault(cls.__name__, cls)
         if len(cls.__bases__) > 0:
             self.wantClass(cls.__bases__[0])
-        return False
-
-    def wantFunction(self, function):
```
Comment (agrawalravi90) [inline test/fw/ptl/utils/plugins/ptl_test_info.py]: was this dead code?

File: test/fw/ptl/utils/plugins/ptl_test_loader.py [hunk 12]
```diff
@@ -165,11 +170,12 @@ def check_unknown(self):
         Check for unknown test suite and test case
         """
         log.debug('check_unknown called')
-        only_ts = self.__tests_list_copy.pop(self._only_ts)
-        only_tc = self.__tests_list_copy.pop(self._only_tc)
+        tests_list_copy = copy.deepcopy(self.__tests_list_copy)
```
Comment (agrawalravi90) [inline test/fw/ptl/utils/plugins/ptl_test_loader.py]: are you modifying the contents of self.__tests_list_copy in this function? I don't see them getting modified, in which case you don't need the deepcopy

File: test/fw/ptl/utils/plugins/ptl_test_loader.py [hunk 13]
```diff
@@ -190,7 +196,19 @@ def prepareTestLoader(self, loader):
         old_loadTestsFromNames = loader.loadTestsFromNames
 
         def check_loadTestsFromNames(names, module=None):
-            rv = old_loadTestsFromNames(names, module)
+            tests_dir = names
+            if not self.testfiles:
+                ptl_test_dir = __file__
+                ptl_test_dir = os.path.join(ptl_test_dir.split('ptl')[0],
```
Comment (agrawalravi90) [inline test/fw/ptl/utils/plugins/ptl_test_loader.py]: don't you also need to modify the Makefiles so that tests get copied to this path when PTL is installed?

File: test/fw/ptl/utils/plugins/ptl_test_loader.py [hunk 14]
```diff
@@ -190,7 +196,19 @@ def prepareTestLoader(self, loader):
...
+                                            "ptl", "tests")
+                user_test_dir = os.environ.get("PTL_TESTS_DIR", None)
+                if user_test_dir and os.path.isdir(user_test_dir):
+                    tests_dir += [user_test_dir]
+                if os.path.isdir(ptl_test_dir):
+                    tests_dir += [ptl_test_dir]
```
Comment (agrawalravi90) [inline test/fw/ptl/utils/plugins/ptl_test_loader.py]: So, this will mean that if PTL_TESTS_DIR is set, then tests will be picked from \<ptl install\>/tests as well as PTL_TESTS_DIR. Is that really what we want? I kind of feel that if users specify PTL_TESTS_DIR, then that should be the ONLY directory where tests should be searched for. What do you think?

File: test/fw/ptl/utils/plugins/ptl_test_loader.py [hunk 15]
```diff
@@ -190,7 +196,19 @@ def prepareTestLoader(self, loader):
...
+                rv = old_loadTestsFromNames(tests_dir, module)
+            else:
+                rv = old_loadTestsFromNames(names, module)
```
Comment (agrawalravi90) [inline test/fw/ptl/utils/plugins/ptl_test_loader.py]: you set tests_dir to name before, so you can just call "rv = old_loadTestsFromNames(tests_dir, module)" right?

File: test/fw/ptl/utils/plugins/ptl_test_loader.py (hunk 13 above)
Comment (shilpakodli) [reply test/fw/ptl/utils/plugins/ptl_test_loader.py]: We are assuming here that when PTL is installed, tests will reside in parallel to PTL library.

File: test/fw/ptl/utils/plugins/ptl_test_loader.py (hunk 12 above)
Comment (shilpakodli) [reply test/fw/ptl/utils/plugins/ptl_test_loader.py]: tests_list_copy.pop removes items. That's why copying self.__tests_list_copy to another variable.

File: test/fw/ptl/utils/plugins/ptl_test_info.py (hunk 11 above)
Comment (shilpakodli) [reply test/fw/ptl/utils/plugins/ptl_test_info.py]: Yes. This was not called anywhere.

Comment (hirenvadalia) [reply test/fw/ptl/utils/plugins/ptl_test_info.py]: @agrawalravi90 It was not the dead code and @shilpakodli it was getting called by nose internal call. And test info plugin was designed such a way that it works without any other ptl node plugin but with this PR changes now we are using loader plugin along with info plugin and loader plugin already has wantFunction so we don't need it here.

File: test/fw/ptl/utils/plugins/ptl_test_info.py (hunk 10 above)
Comment (hirenvadalia) [reply test/fw/ptl/utils/plugins/ptl_test_info.py]: This was there just ensure that we are not adding any class in self._tree which are not subclass of PBSTestSuite (as we want only test class to be added in self._tree) but now that is covered by loader plugin so we don't need to do it here. That's why it was removed

File: test/fw/ptl/utils/plugins/ptl_test_loader.py (hunk 13 above)
Comment (hirenvadalia) [reply test/fw/ptl/utils/plugins/ptl_test_loader.py]: > don't you also need to modify the Makefiles so that tests get copied to this path when PTL is installed?

This is already done. see https://github.com/PBSPro/pbspro/blob/master/test/tests/Makefile.am

File: test/fw/ptl/utils/plugins/ptl_test_loader.py (hunk 14 above)
Comment (hirenvadalia) [reply test/fw/ptl/utils/plugins/ptl_test_loader.py]: We are giving 3 dir to the loader (in a specific order: 1. current dir 2. PTL_TESTS_DIR 3. PTL install dir) to find given tests and loader will load tests from only first dir it find test rest will be ignored for perticular test.

File: test/fw/ptl/utils/plugins/ptl_test_loader.py (hunk 15 above)
Comment (hirenvadalia) [reply test/fw/ptl/utils/plugins/ptl_test_loader.py]: Agree.

File: test/fw/ptl/utils/plugins/ptl_test_loader.py (hunk 14 above)
Comment (agrawalravi90) [reply test/fw/ptl/utils/plugins/ptl_test_loader.py]: Ok, but I feel that it's more intuitive to only look at PTL_TESTS_DIR if it is set by the user. If unset, we can look in current dir + the default install location.

Comment (hirenvadalia) [reply test/fw/ptl/utils/plugins/ptl_test_loader.py]: Ok, but what if the given test is not in PTL_TESTS_DIR but in default install location?

Comment (agrawalravi90) [reply test/fw/ptl/utils/plugins/ptl_test_loader.py]: then we'll say test not found, and that'll be expected behavior. If you disagree then we can ask for an opinion on forums.

Comment (hirenvadalia) [reply test/fw/ptl/utils/plugins/ptl_test_loader.py]: @agrawalravi90 that's what we discussed on forums if I recall it correctly and it concluded that we will give all 3 directly to the loader and let loader find test from one of them and if not found in all 3 then only give test not found error. But saying that I am happy re-discuss that but current code are in line with the current design so if we want to change what you say then first we need to change the design.

Comment (agrawalravi90) [reply test/fw/ptl/utils/plugins/ptl_test_loader.py]: if this has already been discussed then never-mind, I don't want to delay the PR. Let's check it in as is and if somebody else feels that it should change or gets confused by the behavior then we can bring this up again.

**Comment by agrawalravi90** [comment]
@shilpakodli essentially there's only 1 outstanding comment from me, thanks to @hirenvadalia for clarifying other comments. So, you just need to address this one:
"you set tests_dir to name before, so you can just call "rv = old_loadTestsFromNames(tests_dir, module)" right?"

Thanks.

**Comment by shilpakodli** [comment]
Thank you @hiren for clarifying Ravi's questions on behalf of me. Thanks @agrawalravi90 for the review. I have addressed @agrawalravi90 comments.

File: test/fw/ptl/utils/plugins/ptl_test_loader.py [hunk 16]
```diff
@@ -208,7 +208,7 @@ def check_loadTestsFromNames(names, module=None):
                     tests_dir += [ptl_test_dir]
                 rv = old_loadTestsFromNames(tests_dir, module)
             else:
-                rv = old_loadTestsFromNames(names, module)
+                rv = old_loadTestsFromNames(tests_dir, module)
```
Comment (hirenvadalia) [inline test/fw/ptl/utils/plugins/ptl_test_loader.py]: you can remove else part and put old_loadTestsFromNames part outside of if condition
